api/mesh/v1alpha1/gateway.proto

```diff
@@ -0,0 +1,134 @@
+
+import "mesh/options.proto";
+import "mesh/v1alpha1/selector.proto";
+import "system/v1alpha1/datasource.proto";
+import "validate/validate.proto";
+
+// Gateway is a virtual proxy.
+//
+// Each Gateway is bound to a set of builtin gateway dataplanes.
+// Each builtin dataplane instance can host exactly one Gateway
+// proxy configuration.
+//
+// Gateway aligns with the Kubernetes Gateway API v1alpha2. See that
+// spec for detailed documentation.
+message Gateway {
+
+  option (kuma.mesh.resource).name = "GatewayResource";
+  option (kuma.mesh.resource).type = "Gateway";
+  option (kuma.mesh.resource).package = "mesh";
+
+  option (kuma.mesh.resource).skip_registration = true;
+  option (kuma.mesh.resource).skip_validation = true;
+
+  enum Protocol {
+    PROTO_NONE = 0;
+    PROTO_TCP = 1;
+    PROTO_UDP = 2;
+    PROTO_TLS = 3;
+    PROTO_HTTP = 4;
+    PROTO_HTTPS = 5;
+  }
+
+  // TLSConfig describes a TLS configuration.
+  message TLS {
+    enum Mode {
+      // NONE is not a valid TLS mode. Ether TERMINATE or PASSTHRU must
+      // be explicitly configured.
+      TLS_NONE = 0;
+      // The TLS session between the downstream client and the Gateway
+      // is terminated at the Gateway. This mode requires certificateRef
+      // to be set.
+      TLS_TERMINATE = 1;
+      // The TLS session is NOT terminated by the Gateway. This implies
+      // that the Gateway can't decipher the TLS stream except for the
+      // ClientHello message of the TLS protocol. The certificateRef field
+      // is ignored in this mode.
+      TLS_PASSTHRU = 2;
+    }
+
+    message Options {
+      // TODO(jpeach)
+    }
+
+    // Aligns with GatewayTLSConfig.
+    message Conf {
+      // Mode defines the TLS behavior for the TLS session initiated
+      // by the client.
+      Mode mode = 1;
+
+      // CertificateRef is a reference to a object that contains a TLS
+      // certificate and private key. This certificate is used to establish
+      // a TLS session for requests that match the hostname of the associated
+      // listener.
+      kuma.system.v1alpha1.DataSource certificateRef = 2;
+
+      // Options should eventually configure how TLS is configured. This
+      // is where cipher suite and version configuration can be specified,
+      // client certificates enforced, and so on.
+      Options options = 4;
+    }
+  }
+
+  message Listener {
+    // Hostname specifies the virtual hostname to match for protocol types that
+    // define this concept. When unspecified, "", or `*`, all hostnames are
+    // matched. This field can be omitted for protocols that don't require
+    // hostname based matching.
+    string hostname = 1;
+
+    // Port is the network port. Multiple listeners may use the
+    // same port, subject to the Listener compatibility rules.
+    int32 port = 2;
+
+    // Protocol specifies the network protocol this listener expects to receive.
+    Protocol protocol = 3;
```

Comment:
We have `kuma.io/protocol` tag defined on the dataplane's inbound, maybe here we also can use tag instead of separate `protocol` field? But I don't have a strong opinion here, having protocols as `enum` also looks good and explicit

Reply:
Yeh that's a good point. I think that I'm inclined to keep the enums, at least for now since they have well-defined semantics in this context. If we use the mesh tags, then that drags in the mesh semantics which could be confusing (e.g. not sure what we would do with kuma.io/protocol=kafka).